ghcWithPackages environment doesn't build with integer-simple

### Describe the bug

`ghcWithPackages` doesn't appear to work with `integer-simple`.

On 4acff2bbded, I get an error with the following command:

```console
$ nix-shell -I nixpkgs=./. -p 'haskell.packages.integer-simple.ghc8107.ghcWithPackages (p: [p.random])'
this derivation will be built:
  /nix/store/8xw19mncwzyavib9rdwibyz4sh28mljh-ghc-integer-simple-8.10.7-with-packages.drv
building '/nix/store/8xw19mncwzyavib9rdwibyz4sh28mljh-ghc-integer-simple-8.10.7-with-packages.drv'...
/nix/store/7w5cdwwmcccmr5bb8qxayn7969sv743w-random-1.2.0/nix-support:
propagated-build-inputs: /nix/store/7mas5p1hd3xyw8bmgqkkwr1h2jabagpn-splitmix-0.1.0.4/nix-support/propagated-build-inputs
rm: cannot remove '/nix/store/48xd2pxvbi00ph96qr5ba6mwn1ypx9iy-ghc-integer-simple-8.10.7-with-packages/lib/ghc-integer-simple-8.10.7/package.conf.d/package.cache.lock': No such file or direc
tory
```

This command works correctly with non-integer-simple:

```console
$ nix-shell -I nixpkgs=./. -p 'haskell.packages.ghc8107.ghcWithPackages (p: [p.random])'
$ runghc <<< $'import System.Random\nmain = (randomIO :: IO Int) >>= print'
6838749691054437585
```

My guess is that this is caused by the code block in `pkgs/development/haskell-modules/with-packages-wrapper.nix` right here:

https://github.com/NixOS/nixpkgs/blob/18c84ea816348e2a098390101b92d1e39a9dbd45/pkgs/development/haskell-modules/with-packages-wrapper.nix#L132-L140

The `rm` command fails with the error: `rm: cannot remove '/nix/store/48xd2pxvbi00ph96qr5ba6mwn1ypx9iy-ghc-integer-simple-8.10.7-with-packages/lib/ghc-integer-simple-8.10.7/package.conf.d/package.cache.lock': No such file or directory`

This error can be worked around with a change like the following:

```diff
diff --git a/pkgs/development/haskell-modules/with-packages-wrapper.nix b/pkgs/development/haskell-modules/with-packages-wrapper.nix
index e5fe60a0ae5..72cf37f8277 100644
--- a/pkgs/development/haskell-modules/with-packages-wrapper.nix
+++ b/pkgs/development/haskell-modules/with-packages-wrapper.nix
@@ -136,8 +136,8 @@ symlinkJoin {
      # to another nix derivation, so they are not writable.  Removing
      # them allow the correct behavior of ghc-pkg recache
      # See: https://github.com/NixOS/nixpkgs/issues/79441
-     rm $out/lib/${ghc.name}/package.conf.d/package.cache.lock
-     rm $out/lib/${ghc.name}/package.conf.d/package.cache
+     rm ${packageCfgDir}/package.cache.lock
+     rm ${packageCfgDir}/package.cache
 
      $out/bin/${ghcCommand}-pkg recache
      ''}
```

If you look at the definition of `packageCfgDir`, it appears that `ghc.name` is being used where `${ghcCommand}-${ghc.version}` should be used.

However, even with this change, the resulting environment is not correct.  The environment does not correctly contain the `random` package:

```console
$ nix-shell -I nixpkgs=./. -p 'haskell.packages.integer-simple.ghc8107.ghcWithPackages (p: [p.random])'
this derivation will be built:
  /nix/store/5k5fk87zjcdkjz1blrqmf0dvca9vsqwf-ghc-integer-simple-8.10.7-with-packages.drv
building '/nix/store/5k5fk87zjcdkjz1blrqmf0dvca9vsqwf-ghc-integer-simple-8.10.7-with-packages.drv'...
/nix/store/7w5cdwwmcccmr5bb8qxayn7969sv743w-random-1.2.0/nix-support:
propagated-build-inputs: /nix/store/7mas5p1hd3xyw8bmgqkkwr1h2jabagpn-splitmix-0.1.0.4/nix-support/propagated-build-inputs
$ runghc <<< $'import System.Random\nmain = (randomIO :: IO Int) >>= print'

/run/user/1000/runghcXXXX3395958-0.hs:1:1: error:
    Could not find module ‘System.Random’
    Use -v (or `:set -v` in ghci) to see a list of the files searched for.
  |
1 | import System.Random
  | ^^^^^^^^^^^^^^^^^^^^
```

My guess is that this is because the `ghcWithPackages` environment is created with `symlinkJoin`, but the path to the package config directory is different for integer-simple packages and the integer-simple GHC.

Here's what the paths look like for normal GHC:

```console
$ nix-build ./. -A haskell.compiler.ghc8107
/nix/store/ch0a40krsc4zaa9wv8qd1i56pswjzwx0-ghc-8.10.7
$ ls /nix/store/ch0a40krsc4zaa9wv8qd1i56pswjzwx0-ghc-8.10.7/lib/ghc-8.10.7/package.conf.d/
...
$ nix-build ./. -A haskell.packages.ghc8107.random
/nix/store/qamvzkgcl3qcafjc3kngirdvwxhdpqj6-random-1.2.0
$ ls /nix/store/qamvzkgcl3qcafjc3kngirdvwxhdpqj6-random-1.2.0/lib/ghc-8.10.7/package.conf.d/
random-1.2.0-4yTZ53A8d79AtqKspztz45.conf
```

You can see that `random`'s package.conf file ends up in `lib/ghc-8.10.7/package.conf.d/`.

Here's what the paths look like for the integer-simple GHC:

```console
$ nix-build ./. -A haskell.compiler.integer-simple.ghc8107
/nix/store/dbi6hx9fa4lsd3bilq62bqxz0sm41fwb-ghc-integer-simple-8.10.7
$ ls /nix/store/dbi6hx9fa4lsd3bilq62bqxz0sm41fwb-ghc-integer-simple-8.10.7/lib/ghc-8.10.7/package.conf.d/
...
$ nix-build ./. -A haskell.packages.integer-simple.ghc8107.random
/nix/store/7w5cdwwmcccmr5bb8qxayn7969sv743w-random-1.2.0
$ ls /nix/store/7w5cdwwmcccmr5bb8qxayn7969sv743w-random-1.2.0/lib/ghc-integer-simple-8.10.7/package.conf.d/
random-1.2.0-4yTZ53A8d79AtqKspztz45.conf
```

You can see here that `haskell.compiler.integer-simple.ghc8107` has it's `package.conf.d` in `lib/ghc-8.10.7/package.conf.d/`, but `random` has it's `.conf` file in `lib/ghc-integer-simple-8.10.7/package.conf.d/`.  So this `.conf` file doesn't end up in the right place in the `symlinkJoin` in `with-packages-wrapper.nix`.

### Possible Solution

I'm not sure what would be the best solution for this, but I imagine we have to be very careful to make sure GHC and our Nix scripts are looking at the right place for the package conf dir in the following files:

- `pkgs/development/haskell-modules/generic-builder.nix`
- `pkgs/development/haskell-modules/with-packages-wrapper.nix`
- `pkgs/development/compilers/ghc/*.nix`

The `packageCfgDir` variable that is currently defined in `with-packages-wrapper.nix` should likely be an attribute exported by the GHC derivations, and reused in the other files.

### Notify maintainers
@NixOS/haskell 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.81, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.5pre20211126_55275fc`
 - channels(root): `"nixos-21.11.334247.573095944e7"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

--- a/nixos/modules/services/misc/gitit.nix
+++ b/nixos/modules/services/misc/gitit.nix
@@ -10,7 +10,7 @@ let
 
   toYesNo = b: if b then "yes" else "no";
 
-  gititShared = with cfg.haskellPackages; gitit + "/share/" + pkgs.stdenv.hostPlatform.system + "-" + ghc.name + "/" + gitit.pname + "-" + gitit.version;
+  gititShared = with cfg.haskellPackages; gitit + "/share/" + ghc.targetPrefix + ghc.haskellCompilerName + "/" + gitit.pname + "-" + gitit.version;
 
   gititWithPkgs = hsPkgs: extras: hsPkgs.ghcWithPackages (self: with self; [ gitit ] ++ (extras self));
 

--- a/pkgs/development/haskell-modules/configuration-common.nix
+++ b/pkgs/development/haskell-modules/configuration-common.nix
@@ -613,7 +613,7 @@ self: super: {
     doCheck = false;            # https://github.com/kazu-yamamoto/ghc-mod/issues/335
     executableToolDepends = drv.executableToolDepends or [] ++ [pkgs.buildPackages.emacs];
     postInstall = ''
-      local lispdir=( "$data/share/${self.ghc.name}/*/${drv.pname}-${drv.version}/elisp" )
+      local lispdir=( "$data/share/${self.ghc.targetPrefix}${self.ghc.haskellCompilerName}/*/${drv.pname}-${drv.version}/elisp" )
       make -C $lispdir
       mkdir -p $data/share/emacs/site-lisp
       ln -s "$lispdir/"*.el{,c} $data/share/emacs/site-lisp/
@@ -648,7 +648,7 @@ self: super: {
     # cannot easily byte-compile these files, unfortunately, because they
     # depend on a new version of haskell-mode that we don't have yet.
     postInstall = ''
-      local lispdir=( "$data/share/${self.ghc.name}/"*"/${drv.pname}-"*"/elisp" )
+      local lispdir=( "$data/share/${self.ghc.targetPrefix}${self.ghc.haskellCompilerName}/"*"/${drv.pname}-"*"/elisp" )
       mkdir -p $data/share/emacs
       ln -s $lispdir $data/share/emacs/site-lisp
     '';
@@ -659,7 +659,7 @@ self: super: {
     # We cannot easily byte-compile these files, unfortunately, because they
     # depend on a new version of haskell-mode that we don't have yet.
     postInstall = ''
-      local lispdir=( "$data/share/${self.ghc.name}/"*"/${drv.pname}-"*"/elisp" )
+      local lispdir=( "$data/share/${self.ghc.targetPrefix}${self.ghc.haskellCompilerName}/"*"/${drv.pname}-"*"/elisp" )
       mkdir -p $data/share/emacs
       ln -s $lispdir $data/share/emacs/site-lisp
     '';

--- a/pkgs/development/haskell-modules/generic-builder.nix
+++ b/pkgs/development/haskell-modules/generic-builder.nix
@@ -195,13 +195,13 @@ let
     "--prefix=$out"
     "--libdir=\\$prefix/lib/\\$compiler"
     "--libsubdir=\\$abi/\\$libname"
-    (optionalString enableSeparateDataOutput "--datadir=$data/share/${ghc.name}")
+    (optionalString enableSeparateDataOutput "--datadir=$data/share/${ghcNameWithPrefix}")
     (optionalString enableSeparateDocOutput "--docdir=${docdir "$doc"}")
   ] ++ optionals stdenv.hasCC [
     "--with-gcc=$CC" # Clang won't work without that extra information.
   ] ++ [
     "--package-db=$packageConfDir"
-    (optionalString (enableSharedExecutables && stdenv.isLinux) "--ghc-option=-optl=-Wl,-rpath=$out/lib/${ghc.name}/${pname}-${version}")
+    (optionalString (enableSharedExecutables && stdenv.isLinux) "--ghc-option=-optl=-Wl,-rpath=$out/lib/${ghcNameWithPrefix}/${pname}-${version}")
     (optionalString (enableSharedExecutables && stdenv.isDarwin) "--ghc-option=-optl=-Wl,-headerpad_max_install_names")
     (optionalString enableParallelBuilding "--ghc-options=${parallelBuildingFlags}")
     (optionalString useCpphs "--with-cpphs=${cpphs}/bin/cpphs --ghc-options=-cpp --ghc-options=-pgmP${cpphs}/bin/cpphs --ghc-options=-optP--cpp")
@@ -273,6 +273,8 @@ let
 
   ghcCommand' = if isGhcjs then "ghcjs" else "ghc";
   ghcCommand = "${ghc.targetPrefix}${ghcCommand'}";
+
+  ghcNameWithPrefix = "${ghc.targetPrefix}${ghc.haskellCompilerName}";
 
   nativeGhcCommand = "${nativeGhc.targetPrefix}ghc";
 
@@ -349,14 +351,14 @@ stdenv.mkDerivation ({
   # pkgs* arrays defined in stdenv/setup.hs
   + ''
     for p in "''${pkgsBuildBuild[@]}" "''${pkgsBuildHost[@]}" "''${pkgsBuildTarget[@]}"; do
-      ${buildPkgDb nativeGhc.name "$setupPackageConfDir"}
+      ${buildPkgDb "${nativeGhcCommand}-${nativeGhc.version}" "$setupPackageConfDir"}
     done
     ${nativeGhcCommand}-pkg --${nativePackageDbFlag}="$setupPackageConfDir" recache
   ''
   # For normal components
   + ''
     for p in "''${pkgsHostHost[@]}" "''${pkgsHostTarget[@]}"; do
-      ${buildPkgDb ghc.name "$packageConfDir"}
+      ${buildPkgDb ghcNameWithPrefix "$packageConfDir"}
       if [ -d "$p/include" ]; then
         configureFlags+=" --extra-include-dirs=$p/include"
       fi
@@ -493,7 +495,7 @@ stdenv.mkDerivation ({
       # just the target specified; "install" will error here, since not all targets have been built.
     else ''
       ${setupCommand} copy ${buildTarget}
-      local packageConfDir="$out/lib/${ghc.name}/package.conf.d"
+      local packageConfDir="$out/lib/${ghcNameWithPrefix}/package.conf.d"
       local packageConfFile="$packageConfDir/${pname}-${version}.conf"
       mkdir -p "$packageConfDir"
       ${setupCommand} register --gen-pkg-config=$packageConfFile

--- a/pkgs/development/haskell-modules/with-packages-wrapper.nix
+++ b/pkgs/development/haskell-modules/with-packages-wrapper.nix
@@ -51,7 +51,7 @@ let
   ghcCommand = "${ghc.targetPrefix}${ghcCommand'}";
   ghcCommandCaps= lib.toUpper ghcCommand';
   libDir        = if isHaLVM then "$out/lib/HaLVM-${ghc.version}"
-                  else "$out/lib/${ghcCommand}-${ghc.version}";
+                  else "$out/lib/${ghc.targetPrefix}${ghc.haskellCompilerName}";
   docDir        = "$out/share/doc/ghc/html";
   packageCfgDir = "${libDir}/package.conf.d";
   paths         = lib.filter (x: x ? isHaskellLibrary) (lib.closePropagation packages);

--- a/pkgs/development/haskell-modules/with-packages-wrapper.nix
+++ b/pkgs/development/haskell-modules/with-packages-wrapper.nix
@@ -121,7 +121,7 @@ symlinkJoin {
 
   '' + (lib.optionalString (stdenv.targetPlatform.isDarwin && !isGhcjs && !stdenv.targetPlatform.isiOS) ''
     # Work around a linker limit in macOS Sierra (see generic-builder.nix):
-    local packageConfDir="$out/lib/${ghc.name}/package.conf.d";
+    local packageConfDir="${packageCfgDir}";
     local dynamicLinksDir="$out/lib/links"
     mkdir -p $dynamicLinksDir
     # Clean up the old links that may have been (transitively) included by
@@ -148,8 +148,8 @@ symlinkJoin {
      # to another nix derivation, so they are not writable.  Removing
      # them allow the correct behavior of ghc-pkg recache
      # See: https://github.com/NixOS/nixpkgs/issues/79441
-     rm $out/lib/${ghc.name}/package.conf.d/package.cache.lock
-     rm $out/lib/${ghc.name}/package.conf.d/package.cache
+     rm ${packageCfgDir}/package.cache.lock
+     rm ${packageCfgDir}/package.cache
 
      $out/bin/${ghcCommand}-pkg recache
      ''}

